Update Working with Dependencies -> Declaring Dependencies Section (#30507)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/declaring-depedencies/declaring_dependencies_basics.adoc

// Copyright (C) 2023 Gradle, Inc.
//
// Licensed under the Creative Commons Attribution-Noncommercial-ShareAlike 4.0 International License.;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://creativecommons.org/licenses/by-nc-sa/4.0/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[declaring-dependencies-basics]]
= Declaring Dependencies Basics

[[sec:dependency-types]]
== Types of dependencies

> There are three types of dependencies in Gradle:

Author: Done.

1. *Module Dependencies*: Refer to libraries from external repositories.

> 2. *File Dependencies*: Refer to local files or directories, such as `.jar` or `.aar` files.

Author: Done. Re-ordered.

3. *Project Dependencies*: Refer to other projects in the same multi-project build.

[[sub:module_dependencies]]
=== 1. Module dependencies

Module dependencies are the most common dependencies.

> They refer to a module in a repository:

Author: Done.

- **String Notation:** Simplifies dependency declaration by combining the group, name, and version into a single string.
- **Map Notation:** Allows for a more detailed specification, enabling the definition of all properties.

For advanced configurations, such as making a dependency transitive, you can use closures alongside these notations.

Refer to the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html[`DependencyHandler`] class in the API documentation for more examples and details.
Additionally, explore the link:{javadocPath}/org/gradle/api/artifacts/ExternalModuleDependency.html[`ExternalModuleDependency`] class to understand the full range of properties and configuration methods available.

[[sub:file_dependencies]]
=== 2. File dependencies

File dependencies are used when a project relies on external libraries that are not hosted in a binary repository (e.g., JFrog Artifactory or Sonatype Nexus).
Instead, these dependencies might be stored on a shared drive or checked into version control alongside the project source code.

image::dependency-management-file-dependencies.png[]

> File dependencies are unique because they represent a direct reference to files on the filesystem without any associated <<glossary.adoc#sub:terminology_module_metadata,metadata>>, such as transitive dependencies, origin, or author information.

Author: Done.

====

> include::sample[dir="snippets/dependencyManagement/declaringDependencies-fileDependencies/kotlin",files="build.gradle.kts[tags=file-dependencies]"]

Author: It's a lot of text with few examples. So I don't want to remove the snippet. 
However, I think the snippets should be updated in future versions.

====

In this example, each dependency explicitly specifies its location within the file system.
Common methods for referencing these files include:

- link:link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:files(java.lang.Object++[]++)[`Project.files()`]: Accepts one or more file paths directly.
- link:{javadocPath}/org/gradle/api/file/ProjectLayout.html#files-java.lang.Object++...++-[`ProjectLayout.files()`]: Accepts one or more file paths directly.
- link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:fileTree(java.lang.Object)[`Project.fileTree()`]: Defines a directory and includes or excludes specific file patterns.

[NOTE]
====
The order of files in a `FileTree` is not guaranteed to be stable, even on the same computer.
As a result, dependency configurations using `FileTree` may produce resolution results with varying order, which can impact the cacheability of tasks that use these results as inputs.
To ensure more predictable and stable behavior, it is recommended to use the simpler `files` method where possible.
====

> Alternatively, you can use a <<declaring_repositories_adv.adoc#sub:flat_dir_resolver,flat directory>> repository to specify the source directory for multiple file dependencies.

Author: Done.

For more details on working with file dependencies, refer to the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html[`DependencyHandler`] and `ProjectLayout` classes in the Gradle API documentation.

File dependencies also allow you to add a set of files directly to a configuration without using a repository.

> This is useful if you either cannot or choose not to store files in a repository, or if you want to avoid using repositories altogether.

Author: Removed.

Note that file dependencies are not included in the published dependency descriptor for your project.
However, they are available in transitive dependencies within the same build, meaning they can be used within the current build but not outside it.

You can specify which tasks produce the files for a file dependency.
This is useful when files are generated during the build process:

====
include::sample[dir="snippets/artifacts/generatedFileDependencies/kotlin",files="build.gradle.kts[tags=generated-file-dependencies]"]
include::sample[dir="snippets/artifacts/generatedFileDependencies/groovy",files="build.gradle[tags=generated-file-dependencies]"]
====

----
$ gradle -q list
include::{snippetsPath}/artifacts/generatedFileDependencies/tests/generatedFileDependencies.out[]
----

[[sec:versioning_file_dependencies]]
==== Versioning of file dependencies

When using file dependencies, it’s important to clearly express both intent and version.
Unlike module dependencies, file dependencies are not subject to Gradle’s <<dependency_resolution.adoc#sec:how-gradle-downloads-deps,version conflict resolution>>, so assigning a version number in the file name is critical for tracking changes.

> For example, naming a file `commons-beanutils-1.3.jar` helps maintain a clear record of library versions and their associated release notes.

Author: Removed.

[[sub:project_dependencies]]
=== 3. Project dependencies

> Project dependencies allow you to reference other projects or subprojects within a multi-project Gradle build.

Author: Done.

image::dependency-management-project-dependencies.png[]

This is useful for organizing large projects into smaller, modular components:

====
include::sample[dir="snippets/dependencyManagement/declaringDependencies-projectDependencies/kotlin",files="web-service/build.gradle.kts[tags=project-dependencies]"]
include::sample[dir="snippets/dependencyManagement/declaringDependencies-projectDependencies/groovy",files="web-service/build.gradle[tags=project-dependencies]"]
====

Gradle uses the link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:project(java.lang.String)[`project()`] function to define a project dependency.
This function takes the path to the target project within the build.
The path is typically defined using a colon (`:`) to separate different levels of the project hierarchy.

> Project dependencies are automatically included in the correct build order, ensuring that the dependent project is built before the project that depends on it.

Author: Done.

Refer to the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html[`DependencyHandler`] class in the API documentation for more examples and details on how to configure project dependencies within a multi-project build.

[[sec:type-safe-project-accessors]]
==== Type-safe project dependencies

--
Type-safe project accessors are an incubating feature which must be enabled explicitly.
Implementation may change at any time.

> To add support for type-safe project accessors, add `enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")` this to your `settings.gradle(.kts)` file.

Author: Done.

--

One downside of using the `project(":some:path")` notation is the need to remember project paths for dependencies.
Moreover, changing a project path requires manually updating every occurrence, increasing the risk of missing one.

> Since Gradle 7, you can use an experimental type-safe API for project dependencies, which provides IDE completion, making it easier to declare dependencies:

Author: Done.

= Declaring Dependencies Basics

[[sec:dependency-types]]
== Types of dependencies

There are three main types of dependencies in Gradle:

1. *Module Dependencies*: Refer to libraries from external repositories.
2. *Project Dependencies*: Refer to other projects in the same multi-project build.
3. *File Dependencies*: Refer to local files or directories, such as `.jar` or `.aar` files.

[[sub:module_dependencies]]
=== 1. Module dependencies

Module dependencies are the most common dependencies.
They refer to a dependency that is identified by module coordinates (group, name, and version):

====
include::sample[dir="snippets/artifacts/externalDependencies/kotlin",files="build.gradle.kts[tags=module-dependencies]"]
include::sample[dir="snippets/artifacts/externalDependencies/groovy",files="build.gradle[tags=module-dependencies]"]
====

Gradle offers multiple notations for declaring module dependencies, including string notation and map notation.

- **String Notation:** Simplifies dependency declaration by combining the group, name, and version into a single string.
- **Map Notation:** Allows for a more detailed specification, enabling the definition of all properties.

> For advanced configurations, such as making a dependency transitive, you can use closures alongside these notations.

Author: Fixed.

Refer to the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html[`DependencyHandler`] class in the API documentation for more examples and details.

> Additionally, explore the link:{javadocPath}/org/gradle/api/artifacts/ExternalModuleDependency.html[`ExternalModuleDependency`] class to understand the full range of properties and configuration methods available.

Reviewer: 💭 Is there a userguide page that goes into detail on these that would be a better link target?

Author: I removed this since this is the basics and they should not be getting plastered with links so early.

[[sub:project_dependencies]]
=== 2. Project dependencies

Project dependencies allow you to reference other projects within a multi-project Gradle build.

image::dependency-management-project-dependencies.png[]

This is useful for organizing large projects into smaller, modular components:

====
include::sample[dir="snippets/dependencyManagement/declaringDependencies-projectDependencies/kotlin",files="web-service/build.gradle.kts[tags=project-dependencies]"]
include::sample[dir="snippets/dependencyManagement/declaringDependencies-projectDependencies/groovy",files="web-service/build.gradle[tags=project-dependencies]"]
====

Gradle uses the link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:project(java.lang.String)[`project()`] function to define a project dependency.

> This function takes the path to the target project within the build.

Reviewer: ```suggestion
This function takes the relative path to the target project within the build.
```

Author: Fixed.

> The path is typically defined using a colon (`:`) to separate different levels of the project hierarchy.

Reviewer: 🤔 I wouldn't say project hierarchy - the hierarchical relationship between subprojects is kind of thin, and we want to encourage people to think about them independently, for Isolated Projects support.

Author: Updated wording.

To add support for type-safe project accessors, add `enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")` this to your `settings.gradle(.kts)` file:

====
include::sample[dir="snippets/dependencyManagement/declaringDependencies-typeSafeProjectDependencies/kotlin",files="settings.gradle.kts[tags=accessors]"]
include::sample[dir="snippets/dependencyManagement/declaringDependencies-typeSafeProjectDependencies/groovy",files="settings.gradle[tags=accessors]"]
====

One downside of using the `project(":some:path")` notation is the need to remember project paths for dependencies.
Moreover, changing a project path requires manually updating every occurrence, increasing the risk of missing one.

Instead, the experimental type-safe project accessors API provides IDE completion, making it easier to declare dependencies:

====
include::sample[dir="snippets/dependencyManagement/declaringDependencies-typeSafeProjectDependencies/kotlin",files="web-service/build.gradle.kts[tags=project-dependencies]"]
include::sample[dir="snippets/dependencyManagement/declaringDependencies-typeSafeProjectDependencies/groovy",files="web-service/build.gradle[tags=project-dependencies]"]
====

> With this API, incorrect dependencies in Kotlin DSL scripts trigger compilation errors, helping you avoid missed updates.

Reviewer: ```suggestion
With this API, incorrectly specified projects in Kotlin DSL scripts trigger compilation errors, helping you avoid missed updates.
```

[NOTE]
====
The order of files in a `FileTree` is not guaranteed to be stable, even on the same computer.
As a result, dependency configurations using `FileTree` may produce resolution results with varying order, which can impact the cacheability of tasks that use these results as inputs.
To ensure more predictable and stable behavior, it is recommended to use the simpler `files` method where possible.
====

Alternatively, you can use a <<declaring_repositories_adv.adoc#sub:flat_dir_resolver,flat directory>> repository to specify the source directory for multiple file dependencies.

Ideally, you should use Maven or Ivy repository with a local URL:

[source,groovy]
----
repositories {
    maven {
        url 'file:///path/to/local/files' // Replace with your actual path
    }
}
----

For more details on working with file dependencies, refer to the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html[`DependencyHandler`] and `ProjectLayout` classes in the Gradle API documentation.

> File dependencies also allow you to add a set of files directly to a configuration without using a repository.

Reviewer: 💭 This sentence seems like it should go early, perhaps right after "File dependencies allow you to include external JARs or other files directly into your project by referencing their file paths."

Author: Moved.

To add files as dependencies, pass a <<working_with_files.adoc#sec:file_collections,file collection>> to the configuration:

====
include::sample[dir="snippets/artifacts/externalDependencies/kotlin",files="build.gradle.kts[tags=file-dependencies]"]
include::sample[dir="snippets/artifacts/externalDependencies/groovy",files="build.gradle[tags=file-dependencies]"]
====

Note that file dependencies are not included in the published dependency descriptor for your project.
However, they are available in transitive dependencies within the same build, meaning they can be used within the current build but not outside it.

You can specify which tasks produce the files for a file dependency.

> This is useful when files are generated during the build process:

Reviewer: ❗ I think this needs stronger language.  If you don't do this, and depend on these files transitively from another project, the tasks to generate them might not run, and you might not get them in your consumer.

Author: Updated.

This is useful for writing and executing functional tests for Gradle plugins and build scripts:

====
include::sample[dir="snippets/testKit/junitQuickstart/kotlin",files="build.gradle.kts[tags=declare-gradle-testkit-dependency]"]
include::sample[dir="snippets/testKit/junitQuickstart/groovy",files="build.gradle[tags=declare-gradle-testkit-dependency]"]
====

[[sub:groovy_dependencies]]
=== Local Groovy dependency

You can declare a dependency on the Groovy that is distributed with Gradle by using the link:{groovyDslPath}/org.gradle.api.artifacts.dsl.DependencyHandler.html#org.gradle.api.artifacts.dsl.DependencyHandler:localGroovy()[DependencyHandler.localGroovy()] method.
This is useful when you are developing custom Gradle tasks or plugins in Groovy:

====
include::sample[dir="snippets/groovy/groovyDependency/kotlin",files="build.gradle.kts[tags=bundled-groovy-dependency]"]
include::sample[dir="snippets/groovy/groovyDependency/groovy",files="build.gradle[tags=bundled-groovy-dependency]"]
====

[[sec:documenting-dependencies]]
== Documenting dependencies

> When declaring a dependency or a <<dependency_constraints.adoc#dependency-constraints,dependency constraint>>, you can provide a custom reason to clarify why the dependency is included.

Reviewer: ```suggestion
When declaring a dependency or a <<dependency_constraints.adoc#dependency-constraints,dependency constraint>>, you can provide a reason to clarify why the dependency is included.
```

Author: Fixed.
